File: src/hook-output.ts
export function empty(): string {
  return "{}";
}

export function preToolDeny(reason: string, feedback?: string): string {
  return JSON.stringify({
    decision: "block",
    reason,
    hookSpecificOutput: {
      hookEventName: "PreToolUse",
      permissionDecision: "deny",
      permissionDecisionReason: joinFeedback(reason, feedback),
    },
  });
}

export function postToolBlock(reason: string, feedback?: string): string {
  return JSON.stringify({
    decision: "block",
    reason,
    hookSpecificOutput: {
      hookEventName: "PostToolUse",
      additionalContext: joinFeedback(reason, feedback),
    },
  });
}

export function stopBlock(reason: string, feedback?: string): string {
  return JSON.stringify({
    decision: "block",
    reason: joinFeedback(reason, feedback),
  });
}

function joinFeedback(reason: string, feedback?: string): string {
  const trimmed = feedback?.trim();
  if (!trimmed) {
    return reason;
  }
  return `${reason}\n\n${trimmed.slice(-4000)}`;
}

File: markdown_gate/hook_output.py
from __future__ import annotations

import json
from typing import Any


def empty() -> str:
    return "{}"


def pre_tool_deny(reason: str, feedback: str | None = None) -> str:
    payload: dict[str, Any] = {
        "decision": "block",
        "reason": reason,
        "hookSpecificOutput": {
            "hookEventName": "PreToolUse",
            "permissionDecision": "deny",
            "permissionDecisionReason": _join(reason, feedback),
        },
    }
    return json.dumps(payload, ensure_ascii=False)


def post_tool_block(reason: str, feedback: str | None = None) -> str:
    payload: dict[str, Any] = {
        "decision": "block",
        "reason": reason,
        "hookSpecificOutput": {
            "hookEventName": "PostToolUse",
            "additionalContext": _join(reason, feedback),
        },
    }
    return json.dumps(payload, ensure_ascii=False)


def stop_block(reason: str, feedback: str | None = None) -> str:
    payload = {
        "decision": "block",
        "reason": _join(reason, feedback),
    }
    return json.dumps(payload, ensure_ascii=False)


def additional_context(event_name: str, message: str) -> str:
    payload = {
        "hookSpecificOutput": {
            "hookEventName": event_name,
            "additionalContext": message,
        }
    }
    return json.dumps(payload, ensure_ascii=False)


def _join(reason: str, feedback: str | None) -> str:
    if not feedback:
        return reason
    trimmed = feedback.strip()
    if not trimmed:
        return reason
    return f"{reason}\n\n{trimmed[-4000:]}"
